iseq.c

```diff
@@ -248,230 +245,155 @@ rb_iseq_each_value(const rb_iseq_t *iseq, iseq_value_itr_t * func, void *data)
-        rb_iseq_each_value(iseq, update_each_insn_value, (void *)original_iseq);
-
-        if (body->param.flags.has_kw && ISEQ_COMPILE_DATA(iseq) == NULL) {
-            int i, j;
-
-            i = body->param.keyword->required_num;
-
-            for (j = 0; i < body->param.keyword->num; i++, j++) {
-                VALUE obj = body->param.keyword->default_values[j];
-                if (!UNDEF_P(obj)) {
-                    body->param.keyword->default_values[j] = rb_gc_location(obj);
-                }
-            }
-        }
-
-        if (body->catch_table) {
-            struct iseq_catch_table *table = body->catch_table;
-            unsigned int i;
-            for (i = 0; i < table->size; i++) {
-                struct iseq_catch_table_entry *entry;
-                entry = UNALIGNED_MEMBER_PTR(table, entries[i]);
-                if (entry->iseq) {
-                    entry->iseq = (rb_iseq_t *)rb_gc_location((VALUE)entry->iseq);
+                    iseq_scan_bits(i, bits, code, original_iseq);
                 }
             }
         }
-#if USE_MJIT
-        mjit_update_references(iseq);
-#endif
-#if USE_YJIT
-        rb_yjit_iseq_update_references(body->yjit_payload);
-#endif
     }
 }
 
-static VALUE
-each_insn_value(void *ctx, VALUE obj)
-{
-    rb_gc_mark_movable(obj);
-    return obj;
-}
-
 void
-rb_iseq_mark(const rb_iseq_t *iseq)
+rb_iseq_mark_and_update(rb_iseq_t *iseq, bool reference_updating)
 {
     RUBY_MARK_ENTER("iseq");
 
-    RUBY_MARK_UNLESS_NULL(iseq->wrapper);
+    rb_gc_mark_and_move(&iseq->wrapper);
 
     if (ISEQ_BODY(iseq)) {
-        const struct rb_iseq_constant_body *const body = ISEQ_BODY(iseq);
+        struct rb_iseq_constant_body *body = ISEQ_BODY(iseq);
 
-        rb_iseq_each_value(iseq, each_insn_value, NULL);
+        rb_iseq_mark_and_move_each_value(iseq, reference_updating ? ISEQ_ORIGINAL_ISEQ(iseq) : NULL);
 
-        rb_gc_mark_movable(body->variable.coverage);
-        rb_gc_mark_movable(body->variable.pc2branchindex);
-        rb_gc_mark_movable(body->variable.script_lines);
-        rb_gc_mark_movable(body->location.label);
-        rb_gc_mark_movable(body->location.base_label);
-        rb_gc_mark_movable(body->location.pathobj);
-        RUBY_MARK_MOVABLE_UNLESS_NULL((VALUE)body->mandatory_only_iseq);
-        RUBY_MARK_MOVABLE_UNLESS_NULL((VALUE)body->parent_iseq);
+        rb_gc_mark_and_move(&body->variable.coverage);
+        rb_gc_mark_and_move(&body->variable.pc2branchindex);
+        rb_gc_mark_and_move(&body->variable.script_lines);
+        rb_gc_mark_and_move(&body->location.label);
+        rb_gc_mark_and_move(&body->location.base_label);
+        rb_gc_mark_and_move(&body->location.pathobj);
+        if (body->local_iseq) rb_gc_mark_and_move_ptr(&body->local_iseq);
+        if (body->parent_iseq) rb_gc_mark_and_move_ptr(&body->parent_iseq);
+        if (body->mandatory_only_iseq) rb_gc_mark_and_move_ptr(&body->mandatory_only_iseq);
```

Comment:
This portion of code shows the advantage of `rb_gc_mark_and_move`. It can handle most fields, eliminating the need to write two functions, one with `rb_gc_mark_movable` and the other with `rb_gc_location`.